ksql-common/src/main/java/io/confluent/ksql/schema/ksql/SqlType.java

```diff
@@ -19,5 +19,13 @@
  * The SQL types supported by KSQL.
  */
 public enum SqlType {
-  BOOLEAN, INTEGER, BIGINT, DOUBLE, STRING, ARRAY, MAP, STRUCT
+  BOOLEAN, INTEGER, BIGINT, DOUBLE, STRING, ARRAY, MAP, STRUCT;
+
+  public boolean isNumber() {
+    return this == INTEGER || this == BIGINT || this == DOUBLE;
+  }
+
+  public boolean canUpCast(final SqlType to) {
+    return isNumber() && this.ordinal() <= to.ordinal();
```

Comment:
nit: we should try not to rely on `ordinal` since it's a little brittle (aside from the fact that it doesn't really make sense out of numeric context). See javadoc for Enum:
```java
   /**
     * Most programmers will have no use for this method.  It is
     * designed for use by sophisticated enum-based data structures, such
     * as {@link java.util.EnumSet} and {@link java.util.EnumMap}.
```

Instead, I recommend that we manually define the ordering using the `SchemaUtil#ARITHMETIC_TYPE_ORDERING`

Reply:
But that's why we have tests... if someone reorders the tests will fail.  I see no benefit from duplicating the ordinal in a field or adding another data structure.

If this logic was outside of `SqlType`, i.e. where it was before, then I would agree, but I see no issue with the type itself relying on its own ordinal.

Reply:
I think it's slightly clearer to explicitly pass in values for the enum, but if you feel strongly you can drop this. That way you have something like:
```java
BOOLEAN(Integer.MAX_INT),
INTEGER(0),
BIGINT(1),
DOUBLE(2),
STRING(Integer.MAX_INT),
ARRAY(Integer.MAX_INT),
...;

public boolean canUpCast(final SqlType to) {
   return this.numericalHierarchy >= to.numericalHierarchy;
}
```

This is also more extensible. If we add `DECIMAL` which perhaps could be the same hierarchy as `DOUBLE`, we can put them both at the same value.

Reply:
I'd be in favour of leaving it as it is. We can change to something like you're proposing if/when we need to extend it to another comparison type.